src/wp-includes/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -299,25 +299,43 @@ private function process_directives_args( string $html, array &$context_stack, a
 
 			/*
 			 * Sorts the attributes by the order of the `directives_processor` array
-			 * and checks what directives are present in this element. The processing
-			 * order is reversed for tag closers.
+			 * and checks what directives are present in this element.
 			 */
-			$directives_prefixes = array_intersect(
-				$p->is_tag_closer()
-					? $directive_processor_prefixes_reversed
-					: $directive_processor_prefixes,
+			$existing_directives_prefixes          = array_intersect(
+				$directive_processor_prefixes,
 				$directives_prefixes
 			);
+			$existing_directives_prefixes_reversed = array_intersect(
+				$directive_processor_prefixes_reversed,
+				$directives_prefixes
+			);
+			// If it is not a tag closer, process directives in normal order.
```

Comment:
Thinking about it a bit more now that I understand what needs to be done to resolve the issue, I would explore consolidating the code as follows:

```php
			$modes = array(
				'enter' => ! $p->is_tag_closer(),
				'exit'  => $p->is_tag_closer() && ! $p->has_and_visits_its_closer_tag(),
			);
			foreach ( $modes as $mode => $should_run ) {
				if ( ! $should_run ) {
					continue;
				}

				$existing_directives_prefixes = array_intersect(
					$mode === 'enter' ? $directive_processor_prefixes : $directive_processor_prefixes_reversed,
					$directives_prefixes
				);
				foreach ( $existing_directives_prefixes as $directive_prefix ) {
					$func = is_array( self::$directive_processors[ $directive_prefix ] )
						? self::$directive_processors[ $directive_prefix ]
						: array( $this, self::$directive_processors[ $directive_prefix ] );

					call_user_func_array(
						$func,
						array( $p, $mode, &$context_stack, &$namespace_stack, &$tag_stack )
					);
				}
			}
```

This way, the changes applied should be minimal, and void tags should get handled as expected.

In the directive process callbacks, we could pass the `$mode` as illustrated by @dmsnell in another comment. This way the checks there would be easier to follow:

```php
if ( 'exit' === $mode ) {
    // Some logic.
}

Reply:
https://github.com/WordPress/wordpress-develop/pull/6267/commits/01ae2411dc2824820db40a8fc5da865dfde085d1 - exactly what I thought about 👍🏻